### PR TITLE
Misc RCT data integration

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/management/research.c [hunk 1]
```diff
@@ -186,7 +186,7 @@ void research_finish_item(sint32 entryIndex)
 		gResearchedTrackTypesA[base_ride_type] = (RideTypePossibleTrackConfigurations[base_ride_type]         ) & 0xFFFFFFFFULL;
 		gResearchedTrackTypesB[base_ride_type] = (RideTypePossibleTrackConfigurations[base_ride_type] >> 32ULL) & 0xFFFFFFFFULL;
 		if (RideData4[base_ride_type].flags & RIDE_TYPE_FLAG4_3) {
-			ebx = RCT2_GLOBAL(0x0097D4F5 + (base_ride_type * 8), uint8);
+			ebx = RideData4[base_ride_type].var_3;
```
Comment (marijnvdwerf) [inline src/management/research.c]: I don't really get what `var_3` is being used for.

File: src/windows/map.c [hunk 2]
```diff
@@ -91,6 +91,14 @@ static rct_widget window_map_widgets[] = {
 	{ WIDGETS_END },
 };
 
+enum {
+	FLAG_CONSRUCTION_RIGHTS_OWNED = OWNERSHIP_CONSTRUCTION_RIGHTS_OWNED >> 4,
+	FLAG_OWNED = OWNERSHIP_OWNED >> 4,
+	FLAG_CONSTRUCTION_RIGHTS_AVAILABLE = OWNERSHIP_CONSTRUCTION_RIGHTS_AVAILABLE >> 4,
+	FLAG_AVAIALBLE = OWNERSHIP_AVAILABLE >> 4,
+	FLAG_F0 = 0xF0,
```
Comment (marijnvdwerf) [inline src/windows/map.c]: I don't see why this flag is used.

Comment (IntelOrca) [inline src/windows/map.c]: Its a mask.

**Comment by marijnvdwerf** [comment]
But I don't see it set to something other than 1/2/4/8. And that's all within `0x0F`

**Comment by IntelOrca** [comment]
@marijnvdwerf the upper nibble is what fences are shown on the tile.

**Comment by marijnvdwerf** [comment]
@IntelOrca I seriously doubt it. When I follow `GAME_COMMAND_SET_LAND_OWNERSHIP`, the value seems to be just ignored.

**Comment by IntelOrca** [comment]
Sorry, the lower nibble of `sufaceElement->properties.surface.ownership` is the fences and the upper nibble is the `OWNERSHIP_` flags. It uses `0xF0` in `update_park_fences` as a mask to remove all the fences.

**Comment by marijnvdwerf** [comment]
I'm looking at it for the third time now, and I I can only imagine the lower 4 bits being used.

```c
// window_map_tooldrag() { ...
game_do_command(
				gMapSelectPositionA.x,
				GAME_COMMAND_FLAG_APPLY,
				gMapSelectPositionA.y,
				newOwnership, // 0xF6
				GAME_COMMAND_SET_LAND_OWNERSHIP,
				gMapSelectPositionB.x,
				gMapSelectPositionB.y
			);
```

```c
// map_set_land_ownership() { ...
map_buy_land_rights(x1, y1, x2, y2, 6, flags | (newOwnership << 8));
```

```c
// int map_buy_land_rights(int x0, int y0, int x1, int y1, int setting, int flags) { 
cost = map_buy_land_rights_for_tile(x, y, setting, flags);
```
`flags` is `0xF6??` by now

```c
// static money32 map_buy_land_rights_for_tile(int x, int y, int setting, int flags) {
uint8 newOwnership = (flags & 0xFF00) >> 4;
surfaceElement->properties.surface.ownership |= newOwnership;
...
update_park_fences(...)
```
`newOwnership` being an `uint8` would clap `0xF6?` info `0x6?`

**Comment by IntelOrca** [comment]
Look inside `update_park_fences`

**Comment by marijnvdwerf** [comment]
That only uses the '0x6_' part, right? The `F` is lost because of the uint8.

**Comment by IntelOrca** [comment]
This is the format of surfaceElement->properties.surface.ownership:
```
OOOO FFFF
|    |______ Which sides have park fence
--- OWNERSHIP_ flags
```

File: src/management/research.c (hunk 1 above)
Comment (IntelOrca) [inline src/management/research.c]: This code might be wrong, it looks like var_3 represents a track set:
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/ride/track.h#L79

but I think what its indexing into `RideTypePossibleTrackConfigurations` should probably be `base_ride_type `.

**Comment by marijnvdwerf** [comment]
Lets say `newOwnership` gets set to '0xF0 | 0x6'

```c
int flags = GAME_COMMAND_FLAG_APPLY; 
int flags2 = flags | (newOwnership << 8);

uint8 newOwnership2 = (flags & 0xFF00) >> 4;
surfaceElement->properties.surface.ownership |= newOwnership2;
```

The `0xF0` at the start isn't used. `0x6` is used for the ownership part, because of `(0x6 << 8) >> 4`.

**Comment by IntelOrca** [comment]
`0xF0` is used in `update_park_fences`, that is the function you need to look at... forget about flags.

File: src/ride/track.c [hunk 3]
```diff
@@ -818,15 +818,15 @@ static money32 track_place(int rideIndex, int type, int originX, int originY, in
...
+			if (FlatTrackFlags[type] & TRACK_ELEM_FLAG_0200) {
 				if (gTrackGroundFlags & TRACK_ELEMENT_LOCATION_IS_UNDERGROUND) {
 					gGameCommandErrorText = STR_CAN_ONLY_BUILD_THIS_ABOVE_GROUND;
 					return MONEY32_UNDEFINED;
 				}
 			}
 		}
 		else {
-			if (RCT2_ADDRESS(0x0099423C, uint16)[type] & 0x200) {
+			if (TrackFlags[type] & TRACK_ELEM_FLAG_0200) {
```
Comment (Gymnasiast) [inline src/ride/track.c]: Seems to be a flag to mark a piece as only suitable for above ground.

File: src/ride/track.c [hunk 4]
```diff
@@ -835,15 +835,15 @@ static money32 track_place(int rideIndex, int type, int originX, int originY, in
...
+			if (FlatTrackFlags[type] & TRACK_ELEM_FLAG_0001) {
 				if (!(gMapGroundFlags & ELEMENT_IS_UNDERWATER)) {
 					gGameCommandErrorText = STR_CAN_ONLY_BUILD_THIS_UNDERWATER;
 					return MONEY32_UNDEFINED;
 				}
 			}
 		}
 		else {
-			if (RCT2_ADDRESS(0x0099423C, uint16)[type] & 1) {
+			if (TrackFlags[type] & TRACK_ELEM_FLAG_0001) { // No element has this flag
```
Comment (Gymnasiast) [inline src/ride/track.c]: Seems to be a flag to mark a piece as only suitable to be built underwater.

**Comment by janisozaur** [comment]
@marijnvdwerf can you fix AppVeyor?

**Comment by marijnvdwerf** [comment]
@janisozaur I don't really get what's wrong with it

**Comment by janisozaur** [comment]
It seems you included `RideData4` in `S4Importer.cpp`, but `ride_data.h` should have `extern "C"` binding, which you forgot.

File: src/rct1/S4Importer.cpp [hunk 5]
```diff
@@ -24,6 +24,7 @@
 #include "../core/Util.hpp"
 #include "../object/ObjectManager.h"
 #include "Tables.h"
+#include "../ride/ride_data.h"
```
Comment (janisozaur) [inline src/rct1/S4Importer.cpp]: move this to the `extern "C"` block below.

File: src/ride/track_data.c [hunk 6]
```diff
@@ -30699,3 +30699,265 @@ const uint16 TrackFlags[] = {
...
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+};
```
Comment (janisozaur) [inline src/ride/track_data.c]: newline

File: src/ride/track_data.c [hunk 7]
```diff
@@ -30179,3 +30179,785 @@ const rct_vehicle_info_list **gTrackVehicleInfo[17] = {
...
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
+	-1,
```
Comment (janisozaur) [inline src/ride/track_data.c]: 2 too many `-1`.

**Pull request merged**